contracts/utils/structs/Heap.sol

```diff
@@ -0,0 +1,228 @@
+        self.data.pop();
+
+        // ... and heapify
+        _heapifyDown(self, last, 0, lastValue, comp);
+
+        // return root value
+        return rootValue;
+    }
+
+    /**
+     * @dev Insert a new element in the heap using the default comparator.
+     *
+     * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+     * during the lifecycle of a heap will result in undefined behavior.
+     */
+    function insert(Uint256Heap storage self, uint256 value) internal {
+        insert(self, value, Comparators.lt);
+    }
+
+    /**
+     * @dev Insert a new element in the heap using the provided comparator.
+     *
+     * Note: All inserting and removal from a heap should always be done using the same comparator. Mixing comparator
+     * during the lifecycle of a heap will result in undefined behavior.
+     */
+    function insert(
+        Uint256Heap storage self,
+        uint256 value,
+        function(uint256, uint256) view returns (bool) comp
+    ) internal {
+        uint32 size = length(self);
+        self.data.push(Uint256HeapNode({index: size, lookup: size, value: value}));
+        _heapifyUp(self, size, value, comp);
+    }
+
+    /**
+     * @dev Returns the number of elements in the heap.
+     */
+    function length(Uint256Heap storage self) internal view returns (uint32) {
+        return self.data.length.toUint32();
+    }
+
+    function clear(Uint256Heap storage self) internal {
+        Uint256HeapNode[] storage data = self.data;
+        /// @solidity memory-safe-assembly
+        assembly {
+            sstore(data.slot, 0)
+        }
+    }
+
+    /*
+     * @dev Swap node `i` and `j` in the tree.
+     */
+    function _swap(Uint256Heap storage self, uint32 i, uint32 j) private {
+        uint32 ii = self.data[i].index;
+        uint32 jj = self.data[j].index;
+        // update pointers to the data (swap the value)
+        self.data[i].index = jj;
+        self.data[j].index = ii;
+        // update lookup pointers for consistency
+        self.data[ii].lookup = j;
+        self.data[jj].lookup = i;
+    }
+
+    /**
+     * @dev Perform heap maintenance on `self`, starting at position `pos` (with the `value`), using `comp` as a
+     * comparator, and moving toward the leafs of the underlying tree.
+     *
+     * Note: This is a private function that is called in a trusted context with already cached parameters. `lesizength`
+     * and `value` could be extracted from `self` and `pos`, but that would require redundant storage read. These
+     * parameters are not verified. It is the caller role to make sure the parameters are correct.
+     */
+    function _heapifyDown(
+        Uint256Heap storage self,
+        uint32 size,
+        uint32 pos,
+        uint256 value,
+        function(uint256, uint256) view returns (bool) comp
+    ) private {
+        uint32 left = 2 * pos + 1;
+        uint32 right = 2 * pos + 2;
```

Comment:
this could overflow uint32. the checks just after guarantee that they are smaller then size which is uint32.
maybe we could compute as uint256 and then cast down in the ifs.

I don't think its a realistic issue though.